integrators/simple_harmonic_motion.cpp

```diff
@@ -0,0 +1,363 @@
+  parameters_.sampling_period = 1;
+  integrator_->SolveTrivialKineticEnergyIncrement<Length>(
+      ComputeHarmonicOscillatorAcceleration,
+      parameters_, &solution_);
+  Length q_error;
+  Speed v_error;
+  for (std::size_t i = 0; i < solution_.size(); ++i) {
+    q_error = std::max(q_error,
+                       Abs(solution_[i].positions[0].value -
+                           SIUnit<Length>() *
+                           Cos(solution_[i].time.value *
+                               SIUnit<AngularFrequency>())));
+    v_error = std::max(v_error,
+                       Abs(solution_[i].momenta[0].value +
+                           SIUnit<Speed>() *
+                           Sin(solution_[i].time.value *
+                               SIUnit<AngularFrequency>())));
+  }
+  LOG(INFO) << GetParam();
+  LOG(INFO) << "q_error = " << q_error;
+  LOG(INFO) << "v_error = " << v_error;
+#ifdef _DEBUG
+  EXPECT_GE(GetParam().expected_position_error, q_error);
+  EXPECT_GE(GetParam().expected_velocity_error, v_error);
+#else
+  EXPECT_EQ(GetParam().expected_position_error, q_error);
+  EXPECT_EQ(GetParam().expected_velocity_error, v_error);
+#endif
+  // Check consistency with the more general integration schemes.
+  SPRKIntegrator const* const sprk =
+      dynamic_cast<SPRKIntegrator const*>(&*integrator_);
+  if (sprk != nullptr) {
+    SPRKIntegrator::Parameters<Length, Momentum> parameters;
+    parameters.initial.momenta.emplace_back(
+        parameters_.initial.momenta.back().value * Kilogram);
+    parameters.initial.positions = parameters_.initial.positions;
+    parameters.initial.time      = parameters.initial.time;
+    parameters.sampling_period   = parameters_.sampling_period;
+    parameters.tmax              = parameters_.tmax;
+    parameters.tmax_is_exact     = parameters_.tmax_is_exact;
+    parameters.Δt                = parameters_.Δt;
+    SPRKIntegrator::Solution<Length, Momentum> solution;
+    sprk->SolveIncrement<Length, Momentum>(
+        ComputeHarmonicOscillatorForce,
+        ComputeHarmonicOscillatorVelocity,
+        parameters, &solution);
+    Length q_error;
+    Speed v_error;
+    for (std::size_t i = 0; i < solution.size(); ++i) {
+      EXPECT_EQ(solution_[i].momenta.back().value * Kilogram,
+                solution[i].momenta.back().value);
+      EXPECT_EQ(solution_[i].positions.back().value,
+                solution[i].positions.back().value);
+      EXPECT_EQ(solution_[i].time.value, solution[i].time.value);
+    }
+  }
+}
+
+TEST_P(SimpleHarmonicMotionTest, Convergence) {
+  parameters_.initial.positions.emplace_back(SIUnit<Length>());
+  parameters_.initial.momenta.emplace_back(Speed());
+  parameters_.initial.time = Time();
+  parameters_.tmax = 100 * SIUnit<Time>();
+  parameters_.sampling_period = 0;
+  // For 0.2 * 1.1⁻²¹ < |Δt| < 0.2 , the correlation between step size and error
+  // is very strong. It the step is small enough to converge and large enough to
```

Comment:
That whole comment is obsolete now that we have `beginning_of_convergence` actually.